webnn/resources/utils.js

```diff
@@ -1,10 +1,95 @@
 'use strict';
 
-const ExecutionArray = ['async', 'sync'];
-
 // https://webmachinelearning.github.io/webnn/#enumdef-mldevicetype
 const DeviceTypeArray = ['cpu', 'gpu'];
 
+const ExecutionArray = ['async', 'sync'];
+
+// https://webmachinelearning.github.io/webnn/#enumdef-mloperandtype
+// OperandTypeArray was set with supported type list.
+const OperandTypeArray = ['float32'];
+
+const TypedArrayDict = {
+  float32: Float32Array,
+};
+
+/**
+ * Get ULP tolerances of gemm operation.
+ * @param {Array} shapeA [number, number]
+ * @param {Object} options
+ * @param {Object} [options.c]
+ * @param {Number} [options.alpha]
+ * @param {Number} [options.beta]
+ * @param {Boolean} [options.aTranspose]
+ * @param {Boolean} [options.bTranspose]
+ * @returns {Number}
+ */
+function gemmULPTolerances(shapeA, options) {
+  const gemmOptions = {
+    c: 0.0,
+    alpha: 1.0,
+    beta: 1.0,
+    aTranspose: false,
+    bTranspose: false,
+  };
+
+  for (const key in options) {
+    gemmOptions[key] = options[key];
+  }
+
+  const width = gemmOptions.aTranspose ? shapeA[0] : shapeA[1];
```

Comment:
Should we make this relative to the end, so it still works when/if the spec adds leading dimensions (e.g. batch and channel)?

`shapeA[rank(shapeA) - 1]`

Then you can *probably* reuse this logic for matmul too (since the default gemmOptions values would not be overwritten since matmul doesn't have alpha and beta anyway).